File: iOS/Services/CoachOverview.swift
import Foundation

public struct CoachClientSnapshot: Equatable {
    public let id: UUID
    public let name: String

    public init(id: UUID, name: String) {
        self.id = id
        self.name = name
    }
}

public struct CoachRecordSnapshot: Equatable {
    public let date: Date
    public let status: String

    public init(date: Date, status: String) {
        self.date = date
        self.status = status
    }
}

public struct CoachClientSummary: Equatable, Identifiable {
    public var id: UUID { clientId }
    public let clientId: UUID
    public let name: String
    public let takenCount: Int
    public let skippedCount: Int
    public let completionPercent: Int?
    public let lastActivity: Date?
    public let needsCheckIn: Bool
}

public struct CoachTrendPoint: Equatable, Identifiable, Sendable {
    public var id: Date { bucketStart }
    public let bucketStart: Date
    public let takenCount: Int
    public let skippedCount: Int
    public let completionPercent: Int?
}

public struct CoachOverviewSummary: Equatable {
    public let totalClients: Int
    public let activeClients: Int
    public let needsCheckInCount: Int
    public let clients: [CoachClientSummary]
    public let windowDays: Int
    public let takenCount: Int
    public let skippedCount: Int
    public let overallCompletionPercent: Int?
    public let trend: [CoachTrendPoint]
}

public enum CoachOverviewBuilder {
    public static func build(
        clients: [CoachClientSnapshot],
        recordsByClient: [UUID: [CoachRecordSnapshot]],
        now: Date,
        calendar: Calendar = .current,
        windowDays: Int = 7
    ) -> CoachOverviewSummary {
        let safeWindowDays = normalizeWindowDays(windowDays)
        let today = calendar.startOfDay(for: now)
        let start = calendar.date(byAdding: .day, value: -(safeWindowDays - 1), to: today) ?? today
        let end = calendar.date(byAdding: .day, value: 1, to: today) ?? now
        let summaries = clients.map { client in
            buildClientSummary(client, records: recordsByClient[client.id] ?? [], start: start, end: end)
        }.sorted(by: summaryOrder)
        let windowRecords = recordsByClient.values.flatMap { $0 }.filter { $0.date >= start && $0.date < end }
        return makeOverview(summaries, records: windowRecords, start: start, windowDays: safeWindowDays, calendar: calendar)
    }

    private static func makeOverview(
        _ summaries: [CoachClientSummary],
        records: [CoachRecordSnapshot],
        start: Date,
        windowDays: Int,
        calendar: Calendar
    ) -> CoachOverviewSummary {
        let taken = records.filter { $0.status == IntakeStatus.taken.rawValue }.count
        let skipped = records.filter { $0.status == IntakeStatus.skipped.rawValue }.count
        return CoachOverviewSummary(
            totalClients: summaries.count,
            activeClients: summaries.filter { $0.takenCount + $0.skippedCount > 0 }.count,
            needsCheckInCount: summaries.filter { $0.needsCheckIn }.count,
            clients: summaries,
            windowDays: windowDays,
            takenCount: taken,
            skippedCount: skipped,
            overallCompletionPercent: completionPercent(taken: taken, skipped: skipped),
            trend: buildTrend(records: records, start: start, windowDays: windowDays, calendar: calendar)
        )
    }

    private static func buildClientSummary(
        _ client: CoachClientSnapshot,
        records: [CoachRecordSnapshot],
        start: Date,
        end: Date
    ) -> CoachClientSummary {
        let window = records.filter { $0.date >= start && $0.date < end }
        let taken = window.filter { $0.status == IntakeStatus.taken.rawValue }.count
        let skipped = window.filter { $0.status == IntakeStatus.skipped.rawValue }.count
        let completion = completionPercent(taken: taken, skipped: skipped)
        return CoachClientSummary(
            clientId: client.id,
            name: client.name,
            takenCount: taken,
            skippedCount: skipped,
            completionPercent: completion,
            lastActivity: window.map(\.date).max(),
            needsCheckIn: taken + skipped >= 3 && (completion ?? 100) < 70
        )
    }

    private static func buildTrend(
        records: [CoachRecordSnapshot],
        start: Date,
        windowDays: Int,
        calendar: Calendar
    ) -> [CoachTrendPoint] {
        let bucketDays = windowDays <= 7 ? 1 : 7
        let bucketCount = max(1, (windowDays + bucketDays - 1) / bucketDays)
        return (0..<bucketCount).map { index in
            let bucketStart = calendar.date(byAdding: .day, value: index * bucketDays, to: start) ?? start
            let bucketEnd = calendar.date(byAdding: .day, value: bucketDays, to: bucketStart) ?? bucketStart
            return trendPoint(records: records.filter { $0.date >= bucketStart && $0.date < bucketEnd }, start: bucketStart)
        }
    }

    private static func trendPoint(records: [CoachRecordSnapshot], start: Date) -> CoachTrendPoint {
        let taken = records.filter { $0.status == IntakeStatus.taken.rawValue }.count
        let skipped = records.filter { $0.status == IntakeStatus.skipped.rawValue }.count
        return CoachTrendPoint(
            bucketStart: start,
            takenCount: taken,
            skippedCount: skipped,
            completionPercent: completionPercent(taken: taken, skipped: skipped)
        )
    }

    private static func completionPercent(taken: Int, skipped: Int) -> Int? {
        let total = taken + skipped
        return total == 0 ? nil : Int((Double(taken) / Double(total)) * 100)
    }

    private static func normalizeWindowDays(_ windowDays: Int) -> Int {
        [7, 30, 90].contains(windowDays) ? windowDays : 7
    }

    private static func summaryOrder(_ lhs: CoachClientSummary, _ rhs: CoachClientSummary) -> Bool {
        if lhs.needsCheckIn != rhs.needsCheckIn { return lhs.needsCheckIn }
        let lhsCompletion = lhs.completionPercent ?? Int.max
        let rhsCompletion = rhs.completionPercent ?? Int.max
        if lhsCompletion != rhsCompletion { return lhsCompletion < rhsCompletion }
        return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
    }
}

File: Android/app/src/main/java/com/example/supplementtracker/service/CoachOverview.kt
package com.example.supplementtracker.service

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.UUID

data class CoachClientSnapshot(
    val id: UUID,
    val name: String
)

data class CoachRecordSnapshot(
    val epochMs: Long,
    val status: String
)

data class CoachClientSummary(
    val clientId: UUID,
    val name: String,
    val takenCount: Int,
    val skippedCount: Int,
    val completionPercent: Int?,
    val lastActivityEpochMs: Long?,
    val needsCheckIn: Boolean
)

data class CoachOverviewSummary(
    val totalClients: Int,
    val activeClients: Int,
    val needsCheckInCount: Int,
    val clients: List<CoachClientSummary>
)

object CoachOverviewBuilder {
    fun build(
        clients: List<CoachClientSnapshot>,
        recordsByClient: Map<UUID, List<CoachRecordSnapshot>>,
        nowEpochMs: Long,
        zoneId: ZoneId = ZoneId.systemDefault()
    ): CoachOverviewSummary {
        val today = Instant.ofEpochMilli(nowEpochMs).atZone(zoneId).toLocalDate()
        val start = today.minusDays(6)
        val summaries = clients.map { client ->
            buildClientSummary(client, recordsByClient[client.id].orEmpty(), start, today, zoneId)
        }.sortedWith(summaryComparator)
        return CoachOverviewSummary(
            totalClients = summaries.size,
            activeClients = summaries.count { it.takenCount + it.skippedCount > 0 },
            needsCheckInCount = summaries.count { it.needsCheckIn },
            clients = summaries
        )
    }

    private fun buildClientSummary(
        client: CoachClientSnapshot,
        records: List<CoachRecordSnapshot>,
        start: LocalDate,
        today: LocalDate,
        zoneId: ZoneId
    ): CoachClientSummary {
        val window = records.filter { record ->
            val day = Instant.ofEpochMilli(record.epochMs).atZone(zoneId).toLocalDate()
            !day.isBefore(start) && !day.isAfter(today)
        }
        val taken = window.count { it.status == "Taken" }
        val skipped = window.count { it.status == "Skipped" }
        val total = taken + skipped
        val completion = if (total == 0) null else ((taken * 100.0) / total).toInt()
        return CoachClientSummary(
            clientId = client.id,
            name = client.name,
            takenCount = taken,
            skippedCount = skipped,
            completionPercent = completion,
            lastActivityEpochMs = window.maxOfOrNull { it.epochMs },
            needsCheckIn = total >= 3 && (completion ?: 100) < 70
        )
    }

    private val summaryComparator = compareByDescending<CoachClientSummary> { it.needsCheckIn }
        .thenBy { it.completionPercent ?: Int.MAX_VALUE }
        .thenBy { it.name.lowercase() }
}
